File: packages/dashql-core-bindings/test/catalog.test.ts
import '@jest/globals';

import * as dashql from '../src/index.js';
import * as path from 'path';
import * as fs from 'fs';
import { fileURLToPath } from 'node:url';

const distPath = path.resolve(fileURLToPath(new URL('../dist', import.meta.url)));
const wasmPath = path.resolve(distPath, './dashql.wasm');

let dql: dashql.DashQL | null = null;

beforeAll(async () => {
    dql = await dashql.DashQL.create(async (imports: WebAssembly.Imports) => {
        const buf = await fs.promises.readFile(wasmPath);
        return await WebAssembly.instantiate(buf, imports);
    });
    expect(dql).not.toBeNull();
});

describe('Catalog Tests ', () => {
    it('clear catalog', () => {
        const catalog = dql!.createCatalog();
        catalog.addDescriptorPool(1, 10);
        catalog.addSchemaDescriptorT(
            1,
            new dashql.buffers.catalog.SchemaDescriptorT('db1', 'schema1', [
                new dashql.buffers.catalog.SchemaTableT(0, 'table1', [
                    new dashql.buffers.catalog.SchemaTableColumnT('column1'),
                    new dashql.buffers.catalog.SchemaTableColumnT('column2'),
                    new dashql.buffers.catalog.SchemaTableColumnT('column3'),
                ]),
            ]),
        );
        let descriptionBuffer = catalog.describeEntries();
        let description = descriptionBuffer.read();
        expect(description.entriesLength()).toEqual(1);
        descriptionBuffer.destroy();

        descriptionBuffer = catalog.describeEntriesOf(1);
        description = descriptionBuffer.read();
        expect(description.entriesLength()).toEqual(1);
        descriptionBuffer.destroy();

        catalog.clear();

        descriptionBuffer = catalog.describeEntries();
        description = descriptionBuffer.read();
        expect(description.entriesLength()).toEqual(0);
        descriptionBuffer.destroy();

        catalog.destroy();
    });

    it('dynamic registration, one table', () => {
        const catalog = dql!.createCatalog();
        catalog.addDescriptorPool(1, 10);
        expect(catalog.containsEntryId(1)).toBeTruthy();

        // Create and analyze a script referencing an unknown table
        const script = dql!.createScript(catalog, 2);
        script.replaceText('select * from db1.schema1.table1');
        script.scan().destroy();
        script.parse().destroy();
        let analyzedBuffer = script.analyze();
        let analyzed = analyzedBuffer.read();
        expect(analyzed.tableReferencesLength()).toEqual(1);

        // The analyzed script contains an unresolved table ref
        let tableRef = analyzed.tableReferences(0)!;
        expect(tableRef.resolvedTable()).toBeNull();

        // Check the table name
        const tableName = tableRef.tableName(new dashql.buffers.analyzer.QualifiedTableName())!;
        expect(tableName.databaseName()).toEqual('db1');
        expect(tableName.schemaName()).toEqual('schema1');
        expect(tableName.tableName()).toEqual('table1');
        analyzedBuffer.destroy();

        // Resolve the table declaration and add a schema descriptor to the descriptor pool
        catalog.addSchemaDescriptorT(
            1,
            new dashql.buffers.catalog.SchemaDescriptorT('db1', 'schema1', [
                new dashql.buffers.catalog.SchemaTableT(0, 'table1', [
                    new dashql.buffers.catalog.SchemaTableColumnT('column1'),
                    new dashql.buffers.catalog.SchemaTableColumnT('column2'),
                    new dashql.buffers.catalog.SchemaTableColumnT('column3'),
                ]),
            ]),
        );

        // Now analyze the script again
        script.parse().destroy();
        analyzedBuffer = script.analyze();
        analyzed = analyzedBuffer.read();
        expect(analyzed.tableReferencesLength()).toEqual(1);
        tableRef = analyzed.tableReferences(0)!;
        expect(tableRef.resolvedTable()).not.toBeNull();

        // Make sure we set some values for the resolved table
        const resolved = tableRef.resolvedTable(new dashql.buffers.analyzer.ResolvedTable())!;
        expect(resolved.catalogDatabaseId()).not.toEqual(0xFFFFFFFF);
        expect(resolved.catalogSchemaId()).not.toEqual(0xFFFFFFFF);
        expect(dashql.ContextObjectID.isNull(resolved.catalogTableId())).toBeFalsy();

        // Delete all the memory
        analyzedBuffer.destroy();
        script.destroy();
        catalog.destroy();
    });

    it('dynamic registration, multiple tables', () => {
        const catalog = dql!.createCatalog();
        catalog.addDescriptorPool(1, 10);
        expect(catalog.containsEntryId(1)).toBeTruthy();

        // Create and analyze a script referencing an unknown table
        const script = dql!.createScript(catalog, 2);
        script.replaceText('select * from db1.schema1.table1, db1.schema2.table2');
        script.scan().destroy();
        script.parse().destroy();
        let analyzedBuffer = script.analyze();
        let analyzed = analyzedBuffer.read();
        expect(analyzed.tableReferencesLength()).toEqual(2);

        // The analyzed script contains an unresolved table ref
        let tableRef = analyzed.tableReferences(0)!;
        expect(tableRef.resolvedTable()).toBeNull();

        // Check the table name
        const tableName = tableRef.tableName(new dashql.buffers.analyzer.QualifiedTableName())!;
        expect(tableName.databaseName()).toEqual('db1');
        expect(tableName.schemaName()).toEqual('schema1');
        expect(tableName.tableName()).toEqual('table1');
        analyzedBuffer.destroy();

        // Resolve the table declaration and add a schema descriptor to the descriptor pool
        catalog.addSchemaDescriptorsT(
            1,
            new dashql.buffers.catalog.SchemaDescriptorsT([
                new dashql.buffers.catalog.SchemaDescriptorT('db1', 'schema1', [
                    new dashql.buffers.catalog.SchemaTableT(0, 'table1', [
                        new dashql.buffers.catalog.SchemaTableColumnT('column1'),
                        new dashql.buffers.catalog.SchemaTableColumnT('column2'),
                        new dashql.buffers.catalog.SchemaTableColumnT('column3'),
                    ]),
                ]),
                new dashql.buffers.catalog.SchemaDescriptorT('db1', 'schema2', [
                    new dashql.buffers.catalog.SchemaTableT(0, 'table2', [
                        new dashql.buffers.catalog.SchemaTableColumnT('column1'),
                        new dashql.buffers.catalog.SchemaTableColumnT('column2'),
                        new dashql.buffers.catalog.SchemaTableColumnT('column3'),
                    ]),
                ])
            ])
        );

        // Now analyze the script again
        script.parse().destroy();
        analyzedBuffer = script.analyze();
        analyzed = analyzedBuffer.read();
        expect(analyzed.tableReferencesLength()).toEqual(2);
        tableRef = analyzed.tableReferences(0)!;
        expect(tableRef.resolvedTable()).not.toBeNull();

        // Make sure we set some values for the resolved table
        let resolved = tableRef.resolvedTable(new dashql.buffers.analyzer.ResolvedTable())!;
        expect(resolved.catalogDatabaseId()).not.toEqual(0xFFFFFFFF);
        expect(resolved.catalogSchemaId()).not.toEqual(0xFFFFFFFF);
        expect(dashql.ContextObjectID.isNull(resolved.catalogTableId())).toBeFalsy();

        tableRef = analyzed.tableReferences(1)!;
        expect(tableRef.resolvedTable()).not.toBeNull();

        // Make sure we set some values for the second resolved table
        resolved = tableRef.resolvedTable(new dashql.buffers.analyzer.ResolvedTable())!;
        expect(resolved.catalogDatabaseId()).not.toEqual(0xFFFFFFFF);
        expect(resolved.catalogSchemaId()).not.toEqual(0xFFFFFFFF);
        expect(dashql.ContextObjectID.isNull(resolved.catalogTableId())).toBeFalsy();

        // Delete all the memory
        analyzedBuffer.destroy();
        script.destroy();
        catalog.destroy();
    });

    it('tpch flattening', () => {
        const catalog = dql!.createCatalog();

        const schemaScript = dql!.createScript(catalog, 1);
        schemaScript.insertTextAt(0, `
create table part (
   p_partkey integer not null,
   p_name varchar(55) not null,
   p_mfgr char(25) not null,
   p_brand char(10) not null,
   p_type varchar(25) not null,
   p_size integer not null,
   p_container char(10) not null,
   p_retailprice decimal(12,2) not null,
   p_comment varchar(23) not null,
   primary key (p_partkey)
);

create table supplier (
   s_suppkey integer not null,
   s_name char(25) not null,
   s_address varchar(40) not null,
   s_nationkey integer not null,
   s_phone char(15) not null,
   s_acctbal decimal(12,2) not null,
   s_comment varchar(101) not null,
   primary key (s_suppkey)
);

create table partsupp (
   ps_partkey integer not null,
   ps_suppkey integer not null,
   ps_availqty integer not null,
   ps_supplycost decimal(12,2) not null,
   ps_comment varchar(199) not null,
   primary key (ps_partkey,ps_suppkey)
);

create table customer (
   c_custkey integer not null,
   c_name varchar(25) not null,
   c_address varchar(40) not null,
   c_nationkey integer not null,
   c_phone char(15) not null,
   c_acctbal decimal(12,2) not null,
   c_mktsegment char(10) not null,
   c_comment varchar(117) not null,
   primary key (c_custkey)
);

create table orders (
   o_orderkey integer not null,
   o_custkey integer not null,
   o_orderstatus char(1) not null,
   o_totalprice decimal(12,2) not null,
   o_orderdate date not null,
   o_orderpriority char(15) not null,
   o_clerk char(15) not null,
   o_shippriority integer not null,
   o_comment varchar(79) not null,
   primary key (o_orderkey)
);

create table lineitem (
   l_orderkey integer not null,
   l_partkey integer not null,
   l_suppkey integer not null,
   l_linenumber integer not null,
   l_quantity decimal(12,2) not null,
   l_extendedprice decimal(12,2) not null,
   l_discount decimal(12,2) not null,
   l_tax decimal(12,2) not null,
   l_returnflag char(1) not null,
   l_linestatus char(1) not null,
   l_shipdate date not null,
   l_commitdate date not null,
   l_receiptdate date not null,
   l_shipinstruct char(25) not null,
   l_shipmode char(10) not null,
   l_comment varchar(44) not null,
   primary key (l_orderkey,l_linenumber)
);

create table nation (
   n_nationkey integer not null,
   n_name char(25) not null,
   n_regionkey integer not null,
   n_comment varchar(152) not null,
   primary key (n_nationkey)
);

create table region (
   r_regionkey integer not null,
   r_name char(25) not null,
   r_comment varchar(152) not null,
   primary key (r_regionkey)
);
        `);
        schemaScript.scan().destroy();
        schemaScript.parse().destroy();
        schemaScript.analyze().destroy();

        catalog.loadScript(schemaScript, 0);
        expect(catalog.containsEntryId(1)).toBeTruthy();

        const snapPtr = catalog.createSnapshot();
        const snap = snapPtr.read();
        expect(snap.catalogReader.databasesLength()).toEqual(1);
        expect(snap.catalogReader.schemasLength()).toEqual(1);
        expect(snap.catalogReader.tablesLength()).toEqual(8);
        expect(snap.catalogReader.columnsLength()).toEqual(61);

        const db = snap.catalogReader.databases(0)!;
        expect(db.childBegin()).toEqual(0);
        expect(db.childCount()).toEqual(1);
        expect(snap.readName(db.nameId())).toEqual("");

        const schema = snap.catalogReader.schemas(0)!;
        expect(schema.flatParentIdx()).toEqual(0);
        expect(schema.childBegin()).toEqual(0);
        expect(schema.childCount()).toEqual(8);
        expect(snap.readName(schema.nameId())).toEqual("");

        const tableNames = [
            "customer",
            "lineitem",
            "nation",
            "orders",
            "part",
            "partsupp",
            "region",
            "supplier",
        ];

        for (let i = 0; i < 8; ++i) {
            const table = snap.catalogReader.tables(i)!;
            expect(table.flatParentIdx()).toEqual(0);
            expect(snap.readName(table.nameId())).toEqual(tableNames[i]);
        }

        snapPtr.destroy();
        schemaScript.destroy();
        catalog.destroy();
    });
});

File: packages/dashql-core/test/catalog_test.cc
#include "dashql/catalog.h"

#include <flatbuffers/buffer.h>
#include <flatbuffers/flatbuffer_builder.h>

#include <atomic>
#include <chrono>
#include <future>
#include <shared_mutex>
#include <string>
#include <vector>

#include "dashql/analyzer/analyzer.h"
#include "dashql/buffers/index_generated.h"
#include "dashql/catalog_object.h"
#include "dashql/exception.h"
#include "dashql/script.h"
#include "gtest/gtest.h"

using namespace dashql;

namespace {

class LockInspectableCatalog : public Catalog {
   public:
    bool TryLockStateExclusive() {
        if (!state_mutex.try_lock()) {
            return false;
        }
        state_mutex.unlock();
        return true;
    }
};

TEST(CatalogTest, FlattenEmpty) {
    Catalog catalog;
    flatbuffers::FlatBufferBuilder fb;
    fb.Finish(catalog.Flatten(fb));
    auto flat = flatbuffers::GetRoot<buffers::catalog::FlatCatalog>(fb.GetBufferPointer());
    ASSERT_EQ(flat->catalog_version(), catalog.GetVersion());
}

constexpr std::string_view TPCH_SCHEMA = R"SQL(
create table part (
   p_partkey integer not null,
   p_name varchar(55) not null,
   p_mfgr char(25) not null,
   p_brand char(10) not null,
   p_type varchar(25) not null,
   p_size integer not null,
   p_container char(10) not null,
   p_retailprice decimal(12,2) not null,
   p_comment varchar(23) not null,
   primary key (p_partkey)
);

create table supplier (
   s_suppkey integer not null,
   s_name char(25) not null,
   s_address varchar(40) not null,
   s_nationkey integer not null,
   s_phone char(15) not null,
   s_acctbal decimal(12,2) not null,
   s_comment varchar(101) not null,
   primary key (s_suppkey)
);

create table partsupp (
   ps_partkey integer not null,
   ps_suppkey integer not null,
   ps_availqty integer not null,
   ps_supplycost decimal(12,2) not null,
   ps_comment varchar(199) not null,
   primary key (ps_partkey,ps_suppkey)
);

create table customer (
   c_custkey integer not null,
   c_name varchar(25) not null,
   c_address varchar(40) not null,
   c_nationkey integer not null,
   c_phone char(15) not null,
   c_acctbal decimal(12,2) not null,
   c_mktsegment char(10) not null,
   c_comment varchar(117) not null,
   primary key (c_custkey)
);

create table orders (
   o_orderkey integer not null,
   o_custkey integer not null,
   o_orderstatus char(1) not null,
   o_totalprice decimal(12,2) not null,
   o_orderdate date not null,
   o_orderpriority char(15) not null,
   o_clerk char(15) not null,
   o_shippriority integer not null,
   o_comment varchar(79) not null,
   primary key (o_orderkey)
);

create table lineitem (
   l_orderkey integer not null,
   l_partkey integer not null,
   l_suppkey integer not null,
   l_linenumber integer not null,
   l_quantity decimal(12,2) not null,
   l_extendedprice decimal(12,2) not null,
   l_discount decimal(12,2) not null,
   l_tax decimal(12,2) not null,
   l_returnflag char(1) not null,
   l_linestatus char(1) not null,
   l_shipdate date not null,
   l_commitdate date not null,
   l_receiptdate date not null,
   l_shipinstruct char(25) not null,
   l_shipmode char(10) not null,
   l_comment varchar(44) not null,
   primary key (l_orderkey,l_linenumber)
);

create table nation (
   n_nationkey integer not null,
   n_name char(25) not null,
   n_regionkey integer not null,
   n_comment varchar(152) not null,
   primary key (n_nationkey)
);

create table region (
   r_regionkey integer not null,
   r_name char(25) not null,
   r_comment varchar(152) not null,
   primary key (r_regionkey)
);
)SQL";

TEST(CatalogTest, FlattenExampleSchema) {
    Catalog catalog;

    // Create script with TPCH schema
    Script script{catalog};
    script.InsertTextAt(0, TPCH_SCHEMA);
    ASSERT_NO_THROW({
        script.Scan();
        script.Parse();
        script.Analyze();
    });
    auto& analyzed = script.GetAnalyzedScript();

    // Make sure the analyzed script matches expectations
    ASSERT_EQ(analyzed->GetDatabasesByName().size(), 1);
    ASSERT_EQ(analyzed->GetSchemasByName().size(), 1);
    ASSERT_EQ(analyzed->GetTablesByName().size(), 8);

    // Add to catalog
    ASSERT_NO_THROW(catalog.LoadScript(script, 1));

    // Flatten the catalog
    flatbuffers::FlatBufferBuilder fb;
    fb.Finish(catalog.Flatten(fb));
    auto flat = flatbuffers::GetRoot<buffers::catalog::FlatCatalog>(fb.GetBufferPointer());

    // Test the catalog
    ASSERT_EQ(flat->catalog_version(), catalog.GetVersion());
    ASSERT_EQ(flat->databases()->size(), 1);
    ASSERT_EQ(flat->schemas()->size(), 1);
}

TEST(CatalogTest, LoadScriptsPublishesOneRankedGeneration) {
    Catalog catalog;
    Script relations{catalog};
    Script functions{catalog};
    relations.InsertTextAt(0, "create table db.schema.items(id int);");
    functions.InsertTextAt(0, "create function db.schema.item_count() returns int;");
    relations.Analyze();
    functions.Analyze();
    auto previous_version = catalog.GetVersion();

    const Catalog::ScriptBatchEntry batch[] = {{&relations, 20}, {&functions, 10}};
    ASSERT_NO_THROW(catalog.LoadScripts(batch));

    EXPECT_EQ(catalog.GetVersion(), previous_version + 1);
    EXPECT_TRUE(catalog.Contains(relations.GetCatalogEntryId()));
    EXPECT_TRUE(catalog.Contains(functions.GetCatalogEntryId()));
    std::vector<std::tuple<CatalogEntryID, CatalogEntry::Rank, size_t, size_t>> entries;
    catalog.IterateRanked([&](auto id, auto& entry, auto rank) {
        entries.emplace_back(id, rank, entry.GetTables().GetSize(), entry.GetFunctions().GetSize());
    });
    ASSERT_EQ(entries.size(), 2);
    EXPECT_EQ(entries[0], std::make_tuple(functions.GetCatalogEntryId(), 10, 0, 1));
    EXPECT_EQ(entries[1], std::make_tuple(relations.GetCatalogEntryId(), 20, 1, 0));
}

TEST(CatalogTest, LoadScriptsUpdatesLoadedScriptsAndRanksOnce) {
    Catalog catalog;
    Script first{catalog};
    Script second{catalog};
    first.InsertTextAt(0, "create table db.schema.old_table(id int);");
    second.InsertTextAt(0, "create table db.schema.second_table(id int);");
    first.Analyze();
    second.Analyze();
    catalog.LoadScript(first, 1);
    catalog.LoadScript(second, 2);

    first.ReplaceText("create table db.schema.new_table(id int);");
    first.Analyze();
    auto previous_version = catalog.GetVersion();
    const Catalog::ScriptBatchEntry batch[] = {{&first, 30}, {&second, 20}};
    catalog.LoadScripts(batch);

    EXPECT_EQ(catalog.GetVersion(), previous_version + 1);
    std::vector<std::pair<CatalogEntryID, CatalogEntry::Rank>> ranked;
    catalog.IterateRanked([&](auto id, auto&, auto rank) { ranked.emplace_back(id, rank); });
    EXPECT_EQ(ranked, (std::vector<std::pair<CatalogEntryID, CatalogEntry::Rank>>{
                          {second.GetCatalogEntryId(), 20}, {first.GetCatalogEntryId(), 30}}));
    bool found_new_table = false;
    catalog.Iterate([&](auto id, auto& entry) {
        if (id == first.GetCatalogEntryId()) {
            ASSERT_EQ(entry.GetTables().GetSize(), 1);
            found_new_table = entry.GetTables()[0].table_name.table_name.get().text == "new_table";
        }
    });
    EXPECT_TRUE(found_new_table);
}

TEST(CatalogTest, LoadScriptsValidationFailurePreservesCatalog) {
    Catalog catalog;
    Catalog other_catalog;
    Script loaded{catalog};
    Script valid{catalog};
    Script unanalyzed{catalog};
    Script mismatched{other_catalog};
    loaded.InsertTextAt(0, "create table db.schema.loaded(id int);");
    valid.InsertTextAt(0, "create table db.schema.valid(id int);");
    mismatched.InsertTextAt(0, "create table db.schema.other(id int);");
    loaded.Analyze();
    valid.Analyze();
    mismatched.Analyze();
    catalog.LoadScript(loaded, 1);
    auto previous_version = catalog.GetVersion();

    const Catalog::ScriptBatchEntry unanalyzed_batch[] = {{&valid, 2}, {&unanalyzed, 3}};
    EXPECT_THROW(catalog.LoadScripts(unanalyzed_batch), Exception);
    EXPECT_EQ(catalog.GetVersion(), previous_version);
    EXPECT_TRUE(catalog.Contains(loaded.GetCatalogEntryId()));
    EXPECT_FALSE(catalog.Contains(valid.GetCatalogEntryId()));

    const Catalog::ScriptBatchEntry duplicate_batch[] = {{&valid, 2}, {&valid, 3}};
    EXPECT_THROW(catalog.LoadScripts(duplicate_batch), Exception);
    EXPECT_EQ(catalog.GetVersion(), previous_version);
    EXPECT_FALSE(catalog.Contains(valid.GetCatalogEntryId()));

    const Catalog::ScriptBatchEntry mismatch_batch[] = {{&valid, 2}, {&mismatched, 3}};
    EXPECT_THROW(catalog.LoadScripts(mismatch_batch), Exception);
    EXPECT_EQ(catalog.GetVersion(), previous_version);
    EXPECT_FALSE(catalog.Contains(valid.GetCatalogEntryId()));
}

TEST(CatalogTest, WriterWaitsForActiveAnalysis) {
    using namespace std::chrono_literals;

    LockInspectableCatalog catalog;
    Script script{catalog};
    std::string declarations;
    declarations.reserve(250'000);
    for (size_t i = 0; i < 5'000; ++i) {
        declarations += "create table db.schema.table_" + std::to_string(i) + "(a int);";
    }
    script.InsertTextAt(0, declarations);

    std::atomic<bool> analysis_finished = false;
    auto analysis = std::async(std::launch::async, [&] {
        script.Analyze();
        analysis_finished.store(true, std::memory_order_release);
    });

    bool observed_active_analysis = false;
    while (!analysis_finished.load(std::memory_order_acquire)) {
        if (!catalog.TryLockStateExclusive()) {
            observed_active_analysis = true;
            break;
        }
        std::this_thread::yield();
    }
    ASSERT_TRUE(observed_active_analysis);

    auto writer = std::async(std::launch::async, [&] { catalog.Clear(); });
    EXPECT_EQ(writer.wait_for(1ms), std::future_status::timeout);
    analysis.get();
    EXPECT_EQ(writer.wait_for(5s), std::future_status::ready);
    writer.get();
}

}  // namespace
